---
title: "Clustering exploration for `r params$sample_id`"
author: "Maud PLASCHKA"
date: '2024-08-13'
params:
  scpca_project_id: "SCPCP000006"
  sample_id: "SCPCS000176"
  seed: 12345
  padj_thershold: 0.05  # Threshold for adjusted p-value
  lfc_threshold: 1      # Threshold for log fold change
  rate1_threshold: 0.5  # Threshold for rate (percentage) of cells expressing the marker genes in group1
output: 
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
    highlight: pygments
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message=FALSE,
                      warnings=FALSE
                      )
```


## Introduction

The aim is to explore the clustering and label transfers for the sample `r param$sample_id`.

In order to explore the clustering results, we look into some marker genes, pathways enrichment and label transfer.

This approach would provide us a rapid idea of the quality of the clustering:

- in the case each cluster do have a set of specific marker genes, we could expect each of the cluster to have at least a different phenotype,
- if marker genes are mostly shared between clusters, we might have overclustered,
- if no marker genes are found, the quality of cells in the clusters might be impaired (high mitochondrial content? ribosomal content?)

### Packages


```{r packages, message=FALSE, warning=FALSE}
library("Seurat")
library(sctransform)

library(SCpubr)             # for plotting 
library(tidyverse)
library(patchwork)
library(msigdbr)
library(enrichplot)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggalluvial)

set.seed(params$seed)
```


### Base directories

```{r base paths, eval=TRUE}
# The base path for the OpenScPCA repository, found by its (hidden) .git directory
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

# The current data directory, found within the repository base directory
data_dir <- file.path(repository_base, "data", "current", params$scpca_project_id)

# The path to this module
module_base <- file.path(repository_base, "analyses", "cell-type-wilms-tumor-06")
```


### Input files

In this notebook, we are working with the Wilms tumor sample defined in `params$sample_id` from the Wilms tumor dataset `param$scpca_project_id`.
We work with the pre-processed and labeled `Seurat` object that is the output of `02b_label-transfer_fetal_kidney_reference_Stewart.Rmd` saved in the `results` directory.

```{r path_to_query}
result_dir <- file.path(module_base, "results", params$sample_id)
```

To explore the clustering results, we look into some marker genes reported here:

```{r marker_genes, fig.width=12, fig.height=6, out.width='100%'}
CellType_metadata <- read_csv(file.path(module_base, "marker-sets", "CellType_metadata.csv"))

DT::datatable(CellType_metadata, caption = ("CellType_metadata"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',
                               buttons = c( 'csv', 'excel')))
```


### Output file

Reports will be saved in the `notebook` directory. 
The pre-processed and annotated `Seurat` object per samples are saved in the `result` folder.


## Analysis

### Load the pre-processed `Seurat` object
```{r load, message=FALSE, warning=FALSE}
# open the processed rds object
srat <- readRDS(file.path(result_dir, paste0("02b-fetal_kidney_label-transfer_",params$sample_id,".Rds")))
DefaultAssay(srat) <- "SCT"

```


### Functions

Here we defined function that will be used multiple time all along the notebook. 


#### Visualize seurat clusters and markers genes

For a Seurat object `object`and a features `features`, the function `visualize_feature` will plot `FeaturePlot` and `ViolinPlot`

- `object` is the Seurat object 

- `features` the gene or quantitative value to be plotted

- `group.by` is the metadata used for grouping the violin plots 

```{r }
visualize_feature <- function(object, features, group.by = "seurat_clusters"){
 
                  feature_symbol <- AnnotationDbi::select(org.Hs.eg.db, 
                                keys=features, 
                                columns="SYMBOL", 
                                keytype="ENSEMBL")
 
                  d <- SCpubr::do_FeaturePlot(object, 
                                              features = feature_symbol$ENSEMBL, 
                                              pt.size = 0.2, 
                                              legend.width = 0.5, 
                                              legend.length = 5, 
                                              legend.position = "right") + ggtitle(feature_symbol$SYMBOL)
                  b <- SCpubr::do_ViolinPlot(srat, 
                                             features = feature_symbol$ENSEMBL, 
                                             ncol = 1, 
                                             group.by = group.by, 
                                             legend.position = "none",
                                             assay = "SCT") + ylab(feature_symbol$SYMBOL)
                   
                  return(d + b + plot_layout(ncol = 2, widths = c(2,4))) 
}
          
 
```


#### Visualize seurat clusters and metadata

For a Seurat object `object`and a metadata `metadata`, the function `visualize_metadata` will plot `FeaturePlot` and `BarPlot`

- `object` is the Seurat object 

- `metadata` the gene or quantitative value to be plotted

- `group.by` is the metadata used for grouping the violin plots 

```{r }
visualize_metadata <- function(object, meta, group.by){
  
  if(is.numeric(object@meta.data[,meta])){
     d <- SCpubr::do_FeaturePlot(object, 
                                              features = meta, 
                                              pt.size = 0.2, 
                                              legend.width = 0.5, 
                                              legend.length = 5, 
                                              legend.position = "right") + ggtitle(meta)
    b <- SCpubr::do_ViolinPlot(srat, 
                                             features = meta, 
                                             ncol = 1, 
                                             group.by = group.by, 
                                             legend.position = "none")
                   
   return(d + b + plot_layout(ncol = 2, widths = c(2,4))) 
  }
  
  
  else{
    
  
  d <- SCpubr::do_DimPlot(object, reduction="umap", group.by = group.by, label = TRUE, repel = TRUE) + ggtitle(paste0(meta," - umap")) + theme(text=element_text(size=18))
  b <- SCpubr::do_BarPlot(sample = object,
                         group.by = meta,
                         split.by = group.by,
                         position = "fill",
                         font.size = 10,
                         legend.ncol = 3) +
                         ggtitle("% cells")+
                         xlab(print(group.by)) +
                         theme(text=element_text(size=18))
  return(d + b + plot_layout(ncol = 2, widths = c(2,4)))
  }
  
}
```

#### Calculate `ModuleScore` from a `MSigDB` dataset

For a Seurat object `object`, the function `MSigDB_score` will calculate a score using `AddModuleScore()` function for a MSigDB gene set `gs_name` in the category `category`

- `object` is the Seurat object to calculate the score. 
Score will be added in the metadata of this object.

- `category` is a MSigDB collection (https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp).
values in `c("H", "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")`

- `gs_name` is the name of a MSigDB gene set, e.g. `"HALLMARK_P53_PATHWAY"`

- `name` is the name of the module

```{r }
MSigDB_score <- function(object, category, gs_name, name){
  
        set <- msigdbr(species = "human", category = category)
        set_list <- set %>% 
                    dplyr::filter(gs_name == gs_name) %>%
                    dplyr::distinct(gs_name, gene_symbol, human_ensembl_gene) %>% as.data.frame()
        set_list <- list(set_list$human_ensembl_gene)
        object <- AddModuleScore(object, features = set_list, name = name)
return(object)
}
```

#### Enrichment analysis

`Enrichment_plot`aim to perform enrichment of the marker genes for each of the Seurat clusters and summarize the results in a `dotplot`.

- `category` is the MSigDB category or collection to be used
values in `c("H", "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")`

- `signatures` is a list of marker genes per cluster

- `background`is the universe used for enrichment

```{r }
Enrichment_plot <- function(category, signatures, background){
    ## define genesets
   tryCatch( expr =  {
               gene_set <- msigdbr(species = "human", category = category)
               msigdbr_set <-  gene_set %>% dplyr::distinct(gs_name, ensembl_gene) %>% as.data.frame()
               cclust<-compareCluster(geneCluster = signatures, 
               fun = enricher,
               TERM2GENE = msigdbr_set,
               universe=background)
               d <- dotplot(cclust,showCategory=15) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
               return(d)  
   },
   error = function(e) { print("No enrichment") } )
   
}


```

## Analysis content



### Visualize seurat clusters

We expect up to 5 set of clusters:

- blastema cancer cells
- epithelium cancer and/or normal cells
- stroma cancer and/or normal cells
- immune cells
- endothelial cells

```{r fig.width=20, fig.height=10, out.width='100%'}
d2 <- SCpubr::do_DimPlot(srat, reduction="umap", group.by = "seurat_clusters", label = TRUE) + ggtitle("Seurat Cluster - umap")
d1 <- SCpubr::do_DimPlot(srat, reduction="pca", group.by = "seurat_clusters", label = TRUE) + ggtitle("Seurat Cluster - pca")
v1 <- SCpubr::do_ViolinPlot(srat, features = c( "subsets_mito_percent"), ncol = 1, group.by = "seurat_clusters", legend.position = "none") 
v2 <- SCpubr::do_ViolinPlot(srat, features = c( "detected"), ncol = 1, group.by = "seurat_clusters", legend.position = "none") 
v3 <- SCpubr::do_ViolinPlot(srat, features = c( "sum"), ncol = 1, group.by = "seurat_clusters", legend.position = "none") 

d1 + d2 + (v1 + v2 +v3 + plot_layout(ncol=1) ) + plot_layout(ncol = 3, widths = c(2,2,4))
```

### Cell cycle information
```{r cell_cycle}
s.genes <- srat@assays$RNA@meta.data$gene_ids[srat@assays$RNA@meta.data$gene_symbol %in% cc.genes$s.genes]
g2m.genes <- srat@assays$RNA@meta.data$gene_ids[srat@assays$RNA@meta.data$gene_symbol %in% cc.genes$g2m.genes]
srat <- CellCycleScoring(srat, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE)
```

```{r fig.width=10, fig.height=4, out.width='100%'}
visualize_metadata(srat, meta = "Phase", group.by = "seurat_clusters")
```



```{r fig.width=8, fig.height=2.5, out.width='100%'}
visualize_metadata(srat, meta  = "S.Score", group.by = "seurat_clusters")
visualize_metadata(srat, meta = "G2M.Score", group.by = "seurat_clusters")
```

### Look at specific genes

Here, we open the table of marker genes marker-sets/CellType_metadata.csv. 
Note: we do not expect to have a clear and nice pattern of expression for all of the following markers in every tumor. 
This is just ti get a few idea.


```{r fig.width=8, fig.height=2.5, out.width='100%'}
for(feature in CellType_metadata$ENSEMBL_ID[CellType_metadata$ENSEMBL_ID %in% rownames(srat@assays$SCT)]){
   print(visualize_feature(srat, features = feature, group.by = "seurat_clusters"))
}
```



### Look at specific pathways

#### TP53 pathway

here we will calculate a TP53 score using AddMduleScore and the genes of the HALLMARK_P53_PATHWAY gene set.

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "H", gs_name = "HALLMARK_P53_PATHWAY", name = "TP53_score")
visualize_metadata(srat, meta = "TP53_score1", group.by = "seurat_clusters" )
```


#### DNA repair pathway

here we will calculate a DNA_repair score using AddMduleScore and the genes of the HALLMARK_DNA_REPAIR gene set.

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "H", gs_name = "HALLMARK_DNA_REPAIR", name = "DNA_repair_score")
visualize_metadata(srat, meta = "DNA_repair_score1", group.by = "seurat_clusters" )
```

Note: Chemo-treated samples should have higher DNA-damage scores.


#### DROSHA target genes

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "C3", gs_name = "DROSHA_TARGET_GENES", name = "DROSHA_score")
visualize_metadata(srat, meta = "DROSHA_score1", group.by = "seurat_clusters" )
```


#### DICER1 target genes


```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "C3", gs_name = "DICER1_TARGET_GENES", name = "DICER1_score")
visualize_metadata(srat, meta = "DICER1_score1", group.by = "seurat_clusters" )
```


### Find marker genes for each of the seurat clusters

In addition to the list of known marker genes, we used an unbiased approach to find transcripts that characterized the different clusters.
We run DElegate::FindAllMarkers2 to find markers of the different clusters and manually check if they do make sense. 
DElegate::FindAllMarkers2 is an improved version of Seurat::FindAllMarkers based on pseudobulk differential expression method. 
Please check the preprint from Chistoph Hafemeister: https://www.biorxiv.org/content/10.1101/2023.03.28.534443v1
and tool described here: https://github.com/cancerbits/DElegate 

Of note, we won't use it for annotation, this is just here to get an idea!

```{r fig.width=10, fig.height=10, out.width='100%'}
feature_conversion <- srat@assays$RNA@meta.data
de_results   <- DElegate::FindAllMarkers2(srat, group_column = "seurat_clusters")
#filter the most relevant markers
s.markers <- de_results[de_results$padj < params$padj_thershold & de_results$log_fc > params$lfc_threshold & de_results$rate1 > params$rate1_threshold,]
# add gene symbol for easiest interpretation of the result
s.markers$gene_ids <- s.markers$feature
s.markers <- left_join(s.markers,feature_conversion, by = c( "gene_ids") )
stopifnot(
  "Error joining gene ids and feature names" = identical(s.markers$feature, s.markers$gene_ids)   
)
DT::datatable(s.markers, caption = ("marker genes"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',
                               buttons = c( 'csv', 'excel')))
# Select top 5 genes for heatmap plotting
s.markers <- na.omit(s.markers)
s.markers %>%
    group_by(group1) %>%
    top_n(n =  5, wt = log_fc) -> top5
# subset for plotting
cells <- WhichCells(srat, downsample = 100)
ss <- subset(srat, cells = cells)
ss <- ScaleData(ss, features = top5$feature)
p1 <- SCpubr::do_DimPlot(srat, reduction="umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE) + ggtitle("Seurat Cluster - umap")
p2 <- DoHeatmap(ss, features = top5$feature,  cells = cells, group.by = "seurat_clusters") + NoLegend() + 
  scale_fill_gradientn(colors =  c("#01665e","#35978f",'darkslategray3', "#f7f7f7", "#fee391","#fec44f","#F9AD03")) 
p3 <- ggplot(srat@meta.data, aes(seurat_clusters, fill = seurat_clusters)) + geom_bar() + NoLegend()
common_title <- sprintf("Unsupervised clustering %s, %d cells", srat@meta.data$orig.ident[1], ncol(srat))
show((((p1 / p3) + plot_layout(heights = c(3,2)) | p2) ) + plot_layout(widths = c(1, 2)) + plot_layout(heights = c(3,1)) + plot_annotation(title = common_title))
DT::datatable(top5[, c(1, 9, 11, 12)], caption = ("top 5 marker genes"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',
                               buttons = c( 'csv', 'excel')))
```



### Enrichment analysis of marker genes

Here we perform enrichment analysis of the marker genes found in the previous section for each Seurat cluster.

We defined as universe/background all the genes expressed in the dataset, meaning the rownames of the Seurat object.

We used three gene sets from MSiGDB:

- the hallmark gene sets  are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
- the C3 : regulatory target gene sets  based on gene target predictions for microRNA seed sequences and predicted transcription factor binding sites.
- the C8 : cell type signature gene sets  curated from cluster markers identified in single-cell sequencing studies of human tissue.

We used enricher function from clusterProfiler to perform enrichment analysis.

```{r enrichment}
# define background genes = universe for enrichment
background <-  rownames(srat)
# Define gene signature per cluster
signatures <- list()
for( i in unique(s.markers$group1)){
  signatures[[paste0("cluster ",i)]] <- s.markers$feature[s.markers$group1 == i]
}
```

#### Hallmarks MSiGDB gene sets

```{r fig.width=15, fig.height=8, out.width='100%'}

    Enrichment_plot(category = "H", signatures = signatures, background = background)
 
```


EMT signature should be enriched in stroma cluster.
E2F/proliferation should be enriched in blastema cluster.
MYC(N), TP53 must be enriched in blastema cluster.



#### C8 MSiGDB cell type signature gene sets

```{r fig.width=20, fig.height=25, out.width='100%'}
Enrichment_plot(category = "C8", signatures = signatures, background = background)
```

The MSigDB C8 gene set is quite relevant for kidney and nephroblastoma annotations. 
Epithelial (cancer and normal) cells should be enriched in mature/adulte kidney pathways while blastema cancer cells will show enrichment of fetal kidney development pathway / cap mesenchyme. 

### Cell annotation

#### Annotation from SingleR (no cancer or kidney specific dataset)

Here, we quickly checked annotations that are present in the _processed rds object. 
However, the automated annotation have not been performed using a cancer specific reference or a kidney reference. 
We do not expect a nice labelling of the cells as the overlap of cell types between the reference and the query dataset is poor. 
This support the need to do a proper label transfer from the fetal kidney atlas, which is imho the best reference that can be applied to a Wilms tumor query. 


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "singler_celltype_annotation", group.by = "seurat_clusters")
```




#### Annotation from cellassign (no cancer or kidney specific dataset)


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "cellassign_celltype_annotation", group.by = "seurat_clusters")
```



#### Annotation from the fetal reference (Cao et al)


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "fetal_full_predicted.organ", group.by = "seurat_clusters")
visualize_metadata(srat, meta = "fetal_full_predicted.annotation.l1", group.by = "seurat_clusters")
visualize_metadata(srat, meta = "fetal_full_predicted.annotation.l2", group.by = "seurat_clusters")
```



#### Annotation from the kidney reference (Stewart et al)


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "fetal_kidney_predicted.compartment", group.by = "seurat_clusters")
visualize_metadata(srat, meta = "fetal_kidney_predicted.cell_type", group.by = "seurat_clusters")
```

### Compare annotations with marker genes

```{r fig.width=15, fig.height=15, out.width='100%'}

markers = list("malignant" = CellType_metadata$ENSEMBL_ID[CellType_metadata$cell_class == "malignant"],
            "immune"= CellType_metadata$ENSEMBL_ID[CellType_metadata$cell_class == "immune"],
            "endothelium"= CellType_metadata$ENSEMBL_ID[CellType_metadata$cell_class == "endothelium"],
            "non-malignant"= CellType_metadata$ENSEMBL_ID[CellType_metadata$cell_class == "non-malignant"])

p1 <- SCpubr::do_DotPlot(sample = srat,
                         group.by =  "fetal_kidney_predicted.compartment",
                         features = markers, 
                         axis.text.x.angle = 90, 
                         flip = TRUE, 
                         plot.title = "fetal_kidney_predicted.compartment"
                         ) 

p2 <- SCpubr::do_DotPlot(sample = srat,
                         group.by =  "fetal_full_predicted.annotation.l1",
                         features = markers, 
                         axis.text.x.angle = 90, 
                         flip = TRUE,
                         plot.title = "fetal_full_predicted.annotation.l1"
                          )


p3 <- SCpubr::do_DotPlot(sample = srat,
                         group.by =   "seurat_clusters",
                         features = markers, 
                         axis.text.x.angle = 90, 
                         flip = TRUE,
                         plot.title =  "seurat_clusters"
                          )


p4 <- SCpubr::do_DotPlot(sample = srat,
                         group.by =   "cellassign_celltype_annotation",
                         features = markers, 
                         axis.text.x.angle = 90, 
                         flip = TRUE,
                         plot.title = "cellassign_celltype_annotation"
                          )
(p1|p2)/(p3|p4)

```

### Compare annotations between them

```{r fig.width=15, fig.height=15, out.width='100%'}
p1 <- SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "fetal_kidney_predicted.compartment",
                            last_group = "seurat_clusters",
                            repel = TRUE,
                            
                            
                            )

p2 <- SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "fetal_full_predicted.annotation.l1",
                            last_group = "seurat_clusters",
                            repel = TRUE
                            )


p3 <- SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "singler_celltype_annotation",
                            last_group = "seurat_clusters",
                            repel = TRUE
                            )


p4 <- SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "cellassign_celltype_annotation",
                            last_group = "seurat_clusters",
                            repel = TRUE
                            )

(p1|p2)/(p3|p4)
```
```{r fig.width=10, fig.height=15, out.width='100%'}
SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "fetal_kidney_predicted.compartment",
                            last_group = "fetal_full_predicted.annotation.l1",
                            repel = TRUE
                           
                                                        )

SCpubr::do_AlluvialPlot(sample = srat,
                            first_group = "fetal_kidney_predicted.cell_type",
                            last_group = "fetal_full_predicted.annotation.l2",
                            repel = TRUE
                                                        )

```



## Session Info

```{r session info}
# record the versions of the packages used in this analysis and other environment information
sessionInfo()
```


