---
title: "Validation of marker gene expression"
author: Ally Hawkins
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    code_folding: hide
params:
  project_id: "SCPCP000001"
---

This notebook looks at marker gene expression for consensus cell types in `r params$project_id`. 
The marker genes shown in this notebook can be found in [`references/validation-markers.tsv`](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/analyses/cell-type-consensus/references/validation-markers.tsv). 

All plots will also show the final cell type annotation group based on the `validation_group_annotation` in [`references/consensus-validation-groups.tsv`](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/analyses/cell-type-consensus/references/consensus-validation-groups.tsv). 
This means that cells assigned to more granular cell types will be lumped together with a broader cell type, e.g., memory T cell and naive T cell will be renamed as just T cell. 


```{r packages}
suppressPackageStartupMessages({
  # load required packages
  library(ggplot2)
  library(patchwork)
  library(data.table)
})

# Set default ggplot theme
theme_set(
  theme_classic()
)
```


## Data setup


```{r base paths}
# The base path for the OpenScPCA repository, found by its (hidden) .git directory
repository_base <- rprojroot::find_root(rprojroot::is_git_root)
module_base <- file.path(repository_base, "analyses", "cell-type-consensus")

# path to results directory
results_dir <- file.path(repository_base, "data", "current", "results", "cell-type-consensus", params$project_id)
```

```{r}
# marker_heatmap
plotting_functions <- file.path(module_base, "exploratory-notebooks", "utils", "plotting-functions.R")
source(plotting_functions)
```

```{r, results='asis', message=FALSE}
# print out diagnoses for project 
project_diagnoses_file <- file.path(module_base, "sample-info", "project-diagnoses.tsv")
diagnosis <- readr::read_tsv(project_diagnoses_file) |> 
  dplyr::filter(project_id == params$project_id) |> 
  dplyr::pull(diagnosis)

glue::glue("
    <div class=\"alert alert-info\">

    This project contains {diagnosis} samples.

    </div>
  ")
```


```{r, warning=FALSE}
# get consensus results 
consensus_results_files <- list.files(results_dir, pattern = "_processed_consensus-cell-types\\.tsv\\.gz$", recursive = TRUE, full.names = TRUE) 

# get gene expression files 
gene_exp_files <- list.files(results_dir, pattern = "_processed_marker-gene-expression\\.tsv\\.gz$", recursive = TRUE, full.names = TRUE)

# check that consensus results exist
stopifnot(
  "No cell type results found in data/current/results/cell-type-consensus" = length(consensus_results_files) > 0,
  "No gene expression files found in data/current/results/cell-type-consensus" = length(gene_exp_files) > 0
)
```


```{r}
# list of all validation markers
validation_markers_file <- file.path(module_base, "references", "validation-markers.tsv")

# all validation groups mapped to consensus labels 
validation_groups_file <- file.path(module_base, "references", "consensus-validation-groups.tsv")
```


```{r, message=FALSE, warning=FALSE}
# read in consensus and gene expression results 
# use vroom and data.table to try and be more memory-efficient
consensus_dt <- consensus_results_files |> 
  purrr::map(vroom::vroom) |> 
  data.table::rbindlist(fill = TRUE) 

gene_exp_dt <- gene_exp_files |> 
  purrr::map(vroom::vroom) |> 
  data.table::rbindlist(fill = TRUE)

# read in validation markers as data.tables
markers_dt <- fread(validation_markers_file)

validation_groups_dt <- fread(validation_groups_file) |> 
  # rename final assigned group to avoid conflicts when merging in marker gene expression 
  # we want to separate the marker gene group from the actual cell type annotation
  dplyr::select(consensus_annotation, broad_celltype_group = validation_group_annotation)
```


```{r, results='asis'}
# check for if the project is all cell lines or not 
is_cell_line <- all(consensus_dt$sample_type == "cell line")

if(is_cell_line){
  glue::glue("
    <div class=\"alert alert-info\">

    All samples in this project are from cell lines. 
    No cell typing was performed, so no plots will be shown. 

    </div>
  ")
}
```


```{r, eval=!is_cell_line, message=FALSE}
# combine into one data frame that has all marker genes
# one row for each cell/marker gene combination (total rows= total cells x total marker genes)
# includes original consensus annotation, braod cell type group, marker gene id, logcounts, validation group annotation, information regarding expression of marker gene in CellMarker tissues 
consensus_dt <- consensus_dt |> 
  # add in broad cell type group which is used for plotting
  # groups similar cell types together 
  merge(validation_groups_dt, by = "consensus_annotation", all.x = TRUE) |> 
  merge(gene_exp_dt, by = c("barcodes", "library_id"), all.x = TRUE) |> 
  # add marker gene information (associated validation group annotation, gene observed count, percent tissues)
  # account for the same gene being present in multiple cell types 
  merge(markers_dt, by = "ensembl_gene_id", all.x = TRUE, allow.cartesian = TRUE) 

# add a column indicating if gene is present or not
consensus_dt[, detected := logcounts > 0]

# remove the extra data table 
rm(gene_exp_dt)
```

## Plot setup

```{r, eval=!is_cell_line}
# prep for plots 
# get total number of cells per final annotation group 
total_cells_df <- consensus_dt |> 
  dplyr::select(library_id, barcodes, broad_celltype_group) |> 
  unique() |> 
  dplyr::count(broad_celltype_group, name = "total_cells") 
```


```{r, eval=!is_cell_line}
# table with one row per unique broad cell type/ marker gene combination 
# first all cells in with the same broad_celltype_group (determined based on consensus_annotation) are grouped together
# then get the mean gene expression and total percentage of cells that express each marker gene across all cells in that group
# do this for every possible marker gene/ validation group assignment 
# second we calculate the mean expression and mean percentage of all marker genes in a given validation group (this value is used only in the second section of the report)
group_stats_df <- consensus_dt |> 
  # for each assigned cell type/marker gene combo get total detected and mean expression
  # group by both broad group and validation group to account for genes that are expressed in more than one cell type
  dplyr::group_by(broad_celltype_group, ensembl_gene_id, validation_group_annotation) |>
  dplyr::summarize(
    detected_count = sum(detected),
    mean_exp = mean(logcounts)
  ) |> 
  # add in validation group for marker genes
  # this includes all possible marker genes and all possible validation group assignments 
  dplyr::left_join(markers_dt, by = c("ensembl_gene_id", "validation_group_annotation"), relationship = "many-to-many") |>
  # now get the mean expression/ mean percentage across all marker genes for a given validation group
  # here the broad_celltype_group is the final assigned annotation for that group of cells 
  # the validation_group_annotation refers to the cell type that marker gene is associated with 
  dplyr::group_by(broad_celltype_group, validation_group_annotation) |> 
  dplyr::mutate(
    # calculate mean expression/detected across all markers for a specific group 
    all_markers_mean_exp = mean(mean_exp),
    all_markers_detected_count = mean(detected_count)
  ) |>  # add total cells
  dplyr::left_join(total_cells_df, by = c("broad_celltype_group")) |> 
  # for plotting we're only going to look at any cell types with > 50 cells otherwise these plots can get wild 
  dplyr::filter(total_cells > 50) |> 
  dplyr::rowwise() |> 
  dplyr::mutate(
    # get total percent expressed
    percent_exp = (detected_count/total_cells) * 100,
    all_markers_percent_exp = (all_markers_detected_count/total_cells) * 100, 
    # account for NA/unknowns and set axes order
    broad_celltype_group = tidyr::replace_na(broad_celltype_group, "unknown") |> 
      factor(levels =  c(unique(markers_dt$validation_group_annotation), "unknown"))
  ) 

# no longer need this and it takes up space 
rm(consensus_dt)

# make a data frame that just has the unique genes 
# unique genes are those that are only observed in one cell type based on observance in Cell Marker 
# this was determined when creating the marker gene table from Cell Marker 
unique_gene_df <- group_stats_df |> 
  # keep all 6 HPC genes 
  dplyr::filter(gene_observed_count == 1 | validation_group_annotation == "hematopoietic precursor cell")
```

```{r, eval=!is_cell_line}
# get list of celltypes to keep and assign colors 
celltype_groups <- group_stats_df |> 
  dplyr::pull(broad_celltype_group) |> 
  unique() |>
  as.character()

# assign colors to use in annotation bars 
celltype_colors <- palette.colors(palette = "alphabet")[1:length(celltype_groups)]
names(celltype_colors) <- celltype_groups
```


## Expression of unique marker genes 

First we will look at the expression of marker genes that are specific to a cell type group.  
All marker genes shown in the following plots are only found to be a marker gene for a single cell type group with the exception of marker genes for hematopoietic precursor cells, as none of those genes are unique to a single cell type and may also be found in other immune cells.
Only cell type groups with at least 50 cells will be shown. 

The dot plot shows the mean expression of each marker gene across all cells assigned to a given cell type group. 
Only genes that have mean expression > 0 and are expressed in at least 5% of cells in at least one cell type group will be shown. 

```{r, eval=!is_cell_line}
# filter markers to those that are actually relevant 
# we will only plot the marker genes for cell types that are part of the assigned broad validation group for this project
# we don't care about plotting marker genes for cell types that aren't present here 
# note that we will use this for both the dotplot and the heatmap
filtered_markers_df <- markers_dt |> 
  dplyr::filter(validation_group_annotation %in% celltype_groups,
                gene_symbol %in% group_stats_df$gene_symbol)

# now only keep unique markers for the dotplot 
unique_markers_df <- filtered_markers_df |> 
  dplyr::filter(gene_observed_count == 1 | validation_group_annotation == "hematopoietic precursor cell")

# specify x axis order for dotplot
marker_gene_order <- unique_markers_df |> 
  dplyr::pull(gene_symbol)

# set order for cell types 
celltype_order <- unique(unique_markers_df$validation_group_annotation)
```

```{r, eval=!is_cell_line, fig.height=10, fig.width=15}
# filter out low expressed genes
dotplot_df <- unique_gene_df |> 
  dplyr::filter(mean_exp > 0, percent_exp > 5) |> 
  dplyr::arrange(broad_celltype_group) |> 
  # add a label for the plot 
  dplyr::mutate(y_label = as.factor(glue::glue("{broad_celltype_group} ({total_cells})"))) |> 
  # remove marker genes that aren't present in final annotations and set x axis order 
  dplyr::filter(gene_symbol %in% marker_gene_order) |> 
  dplyr::mutate(
    # set orders of gene symbol and validation groups 
    gene_symbol = factor(gene_symbol, levels = marker_gene_order),
    validation_group_annotation = factor(validation_group_annotation, levels = celltype_order)
  )

# make dotplot with marker gene exp
dotplot <- ggplot(dotplot_df, aes(y = forcats::fct_rev(y_label), x = gene_symbol, color = mean_exp, size = percent_exp)) +
  geom_point() +
  scale_color_viridis_c(option = "magma") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  ) +
  labs(
    x = "",
    y = "Broad cell type annotation",
    color = "Mean gene expression",
    size = "Percent cells expressed"
  )


# add annotation bar aligning marker genes with validation group 
color_bar <- ggplot(dotplot_df, aes(x = gene_symbol, y = 1, fill = validation_group_annotation)) + 
  geom_tile() + 
  scale_fill_manual(values = celltype_colors, breaks = levels(dotplot_df$validation_group_annotation)) +
  ggmap::theme_nothing() +
  theme(legend.position = "bottom") +
  labs(fill = "")

dotplot + color_bar +
  patchwork::plot_layout(ncol = 1, heights = c(4, 0.1)) 
```


## Mean expression across all marker genes 

We will make the same dot plot but instead of looking at the expression of individual marker genes for all cell type groups, we will plot the mean expression of all marker genes for a specific cell type group. 

This is calculated by first getting the mean expression of each individual marker gene across all cells in a cell type group and then taking the mean of that value across all marker genes. 
This results in one row per assigned cell type group and one column per marker gene group. 

```{r, eval=!is_cell_line}
# filter low expressed genes 
grouped_dotplot_df <- group_stats_df |> 
  dplyr::select(broad_celltype_group, validation_group_annotation, all_markers_mean_exp, all_markers_percent_exp, total_cells) |>
  unique() |> 
  dplyr::filter(all_markers_mean_exp > 0, all_markers_percent_exp > 5, validation_group_annotation %in% celltype_groups) |>
  dplyr::arrange(broad_celltype_group) |> 
  # make sure labels are correct 
  dplyr::mutate(y_label = as.factor(glue::glue("{broad_celltype_group} ({total_cells})")),
                validation_group_annotation = factor(validation_group_annotation, levels = celltype_order))


ggplot(grouped_dotplot_df, aes(y = forcats::fct_rev(y_label), x = validation_group_annotation, color = all_markers_mean_exp, size = all_markers_percent_exp)) +
  geom_point() +
  scale_color_viridis_c(option = "magma") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  ) +
  labs(
    x = "Marker gene group",
    y = "Broad cell type annotation", 
    color = "Mean gene expression", 
    size = "Percent cells expressed"
  )

```

## Session info 

```{r session info}
# record the versions of the packages used in this analysis and other environment information
sessionInfo()
```
