---
title: "Exploration of clustering results for `r params$sample_id`"
author: maud-p
date: '`r format(Sys.time(), "%B %d, %Y %H:%M:%S %Z")`'
output:
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
    highlight: pygments
    df_print: kable
    self_contained: yes
    mode: selfcontained 
params:
  current: "2024-07-08"
  scpca_project_id: "SCPCP000006"
  sample_id: "SCPCS000176"
  seed: 12345
  padj_thershold: 0.05  # Threshold for adjusted p-value
  lfc_threshold: 1      # Threshold for log fold change
  rate1_threshold: 0.5  # Threshold for rate (percentage) of cells expressing the marker genes in group1
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warnings = FALSE
)
```


## Introduction

This a template for an analysis notebook using `RMarkdown`.


In this notebook, we will perform the `Seurat` workflow [normalization --> reduction --> clustering] and explore the clustering results. 

This correspond to the step 2 of the proposed analysis: clustering of cells across a set of parameters for few samples

- https://github.com/AlexsLemonade/OpenScPCA-analysis/discussions/635
- https://github.com/AlexsLemonade/OpenScPCA-analysis/issues/671

In order to assess the clustering quality, we look into some marker genes, pathways enrichment and label transfer.
This approach would provide us a rapid idea of the quality of the clustering:

- in the case each cluster do have a set of specific marker genes, we could expect each of the cluster to have at least a different phenotype,
- if marker genes are mostly shared between clusters, we might have overclustered,
- if no marker genes are found, the quality of cells in the clusters might be impaired (high mitochondrial content? ribosomal content?)


### Description of the analysis workflow

We perform the following analysis to assess for the quality of clustering:

[1] We perform some quality check to assess any QC-induced clustering (nFeature, nCount, percent.mito). 

[2] We add cell cycle information, as we know that in a specific cell cycle state, the transcriptional program is mostly/exclusively related to cell cycle genes and the identity of cells is difficult to determine. 
We expect these cells to cluster together in a cluster of proliferating cells. 

[3] We look at specific marker genes that we reported in the table marker.sets/CellType_metadata.csv to check the relevance of the clustering.

[4] We look at specific pathways to check the relevance of the clustering.

[5] We run `DElegate::FindAllMarkers2` to find markers of the different clusters and manually check if they do make sense. 
`DElegate::FindAllMarkers2` is an improved version of `Seurat::FindAllMarkers` based on pseudobulk differential expression method. 

[6] We perform enrichment analysis of marker genes for each Seurat clusters. 
We defined all the genes from the Seurat object as the universe and used the MSigDB gene sets. 

[7] We plot `pca`/`umap` reduction grouping with available annotations (`singler_`, `cellassign_`). 
We expect at least immune cells to be correctly label and fall into a few set of clusters.

[ ] Next step, aim for a next PR: We will transfer annotations from two human fetal (kidney) references (Stewart et al., Cao et al.). 
We plot pca/umap reduction grouping with latest labels. 
We expect it to be the most representative of the cell types in the sample. 

### Packages

Load required packages in the following chunk, if needed.
Do not install packages here; only load them with the `library()` function.

```{r packages, message=FALSE, warning=FALSE}
library(SingleCellExperiment)

library("Seurat")

#library(Azimuth)            # Will be required once we add the labl transfer from Azimuth fetal reference
#library(SeuratData)         # Will be required once we add the labl transfer from Azimuth fetal reference

library(viridis)            # for plotting with viridis colors
library(SCpubr)             # for plotting 
library(tidyverse)
library(patchwork)

library(msigdbr)
library(enrichplot)
library(clusterProfiler)

library(org.Hs.eg.db)


set.seed(params$seed)
```

### Base directories

```{r base paths}
# The base path for the OpenScPCA repository, found by its (hidden) .git directory
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

# The current data directory, found within the repository base directory
data_dir <- file.path(repository_base, "data", params$current, params$scpca_project_id)

# The path to this module
module_base <- file.path(repository_base, "analyses", "cell-type-wilms-tumor-06")
```

### Selected parameters

Please note: to keep the notebook as straight as possible, we decided to show the analysis for the selected set of parameters:

- `SCTransform` (default parameters)
- `RunPCA` (default) 
- `RunUMAP` (dims = 1:50) 
- `FindNeighbors` (dims = 1:50) 
- `FindClusters` (default)


Note: Other parameters have been previously tested, but we would like to show in the following report that the one selected is performing good.

### Functions

Here we defined function that will be used multiple time all along the notebook. 


#### Visualize seurat clusters and markers genes

For a Seurat object `object`and a features `features`, the function `visualize_feature` will plot `FeaturePlot` and `ViolinPlot`

- `object` is the Seurat object 

- `features` the gene or quantitative value to be plotted

- `group.by` is the metadata used for grouping the violin plots 

```{r }

visualize_feature <- function(object, features, group.by = "seurat_clusters"){
 
                  feature_symbol = AnnotationDbi::select(org.Hs.eg.db, 
                                keys=features, 
                                columns="SYMBOL", 
                                keytype="ENSEMBL")

 
                  d <- SCpubr::do_FeaturePlot(object, 
                                              features = feature_symbol$ENSEMBL, 
                                              pt.size = 0.2, 
                                              legend.width = 0.5, 
                                              legend.length = 5, 
                                              legend.position = "right") + ggtitle(feature_symbol$SYMBOL)

                  b <- SCpubr::do_ViolinPlot(srat, 
                                             features = feature_symbol$ENSEMBL, 
                                             ncol = 1, 
                                             group.by = group.by, 
                                             legend.position = "none") + ylab(feature_symbol$SYMBOL)
                   
                  return(d + b + plot_layout(ncol = 2, widths = c(2,4))) 

}
          
 
```


#### Visualize seurat clusters and metadata

For a Seurat object `object`and a metadata `metadata`, the function `visualize_metadata` will plot `FeaturePlot` and `BarPlot`

- `object` is the Seurat object 

- `metadata` the gene or quantitative value to be plotted

- `group.by` is the metadata used for grouping the violin plots 

```{r }

visualize_metadata <- function(object, meta, group.by){
  
  if(class(object@meta.data[,meta]) == "numeric"){
     d <- SCpubr::do_FeaturePlot(object, 
                                              features = meta, 
                                              pt.size = 0.2, 
                                              legend.width = 0.5, 
                                              legend.length = 5, 
                                              legend.position = "right") + ggtitle(meta)

    b <- SCpubr::do_ViolinPlot(srat, 
                                             features = meta, 
                                             ncol = 1, 
                                             group.by = group.by, 
                                             legend.position = "none")
                   
   return(d + b + plot_layout(ncol = 2, widths = c(2,4))) 
  }
  
  
  else{
    
  
  d <- SCpubr::do_DimPlot(object, reduction="umap", group.by = group.by, label = TRUE, repel = TRUE) + ggtitle(paste0(meta," - umap")) + theme(text=element_text(size=18))

  b <- SCpubr::do_BarPlot(sample = object,
                         group.by = meta,
                         split.by = group.by,
                         position = "fill",
                         font.size = 10,
                         legend.ncol = 3) +
                         ggtitle("% cells")+
                         xlab(file) +
                         theme(text=element_text(size=18))
  return(d + b + plot_layout(ncol = 2, widths = c(2,4)))
  }
  
}

```

#### Calculate `ModuleScore` from a `MSigDB` dataset

For a Seurat object `object`, the function `MSigDB_score` will calculate a score using `AddModuleScore()` function for a MSigDB gene set `gs_name` in the category `category`

- `object` is the Seurat object to calculate the score. 
Score will be added in the metadata of this object.

- `category` is a MSigDB collection (https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp).
values in `c("H", "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")`

- `gs_name` is the name of a MSigDB gene set, e.g. `"HALLMARK_P53_PATHWAY"`

- `name` is the name of the module

```{r }

MSigDB_score <- function(object, category, gs_name, name){
  
        set <- msigdbr(species = "human", category = category)

        set_list <- set %>% 
                    dplyr::filter(gs_name == gs_name) %>%
                    dplyr::distinct(gs_name, gene_symbol, human_ensembl_gene) %>% as.data.frame()

        set_list <- list(set_list$human_ensembl_gene)

        object <- AddModuleScore(object, features = set_list, name = name)

return(object)
}

```

#### Enrichment analysis

`Enrichment_plot`aim to perform enrichment of the marker genes for each of the Seurat clusters and summarize the results in a `dotplot`.

- `category` is the MSigDB category or collection to be used
values in `c("H", "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")`

- `signatures` is a list of marker genes per cluster

- `background`is the universe used for enrichment

```{r }

Enrichment_plot <- function(category, signatures, background){

    ## define genesets
    gene_set <- msigdbr(species = "human", category = category)
    msigdbr_set <-  gene_set %>% dplyr::distinct(gs_name, ensembl_gene) %>% as.data.frame()

    cclust<-compareCluster(geneCluster = signatures, 
               fun = enricher,
               TERM2GENE = msigdbr_set,
               universe=background)
    d <- dotplot(cclust,showCategory=15) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    return(d)  
}

```

## Analysis content

### Load the processed SingleCellExperiment rds object 


```{r load}
filelist <- list.files(data_dir, full.names = TRUE)

# select the 40 Wilms tumor single nucleus folders only
filedir <- filelist[grepl(params$sample_id, filelist)]

file <- list.files(filedir)
file <- file[grepl("_processed.rds", file)]

# open the processed rds object
sce <- readRDS(paste0(filedir, "/", file))
```


### Run Seurat pipeline

```{r seurat_wf}
# convert to seurat
srat <- CreateSeuratObject(counts = counts(sce),
                                    assay = "RNA",
                                    project = params$sample_id
                           )

# convert colData and rowData to data.frame for use in the Seurat object
cell_metadata <- as.data.frame(colData(sce))
row_metadata <- as.data.frame(rowData(sce))

# add cell metadata (colData) from SingleCellExperiment to Seurat
srat@meta.data <- cell_metadata

# add row metadata (rowData) from SingleCellExperiment to Seurat
srat[["RNA"]]@meta.data <- row_metadata

# add metadata from SingleCellExperiment to Seurat
srat@misc <- metadata(sce)
rm(sce)

# Normalization
srat <- SCTransform(srat, verbose = F, conserve.memory = TRUE)

# dimensionality reduction
srat <- RunPCA(srat, verbose = F)
srat <- RunUMAP(srat, dims = 1:50, verbose = F)

# clustering
srat    <- FindNeighbors(srat, dims = 1:50, verbose = F)
srat    <- FindClusters(srat, verbose = T)
  
```



### Visualize seurat clusters

```{r fig.width=20, fig.height=5, out.width='100%'}

d2 <- SCpubr::do_DimPlot(srat, reduction="umap", group.by = "seurat_clusters", label = TRUE) + ggtitle("Seurat Cluster - umap")
d1 <- SCpubr::do_DimPlot(srat, reduction="pca", group.by = "seurat_clusters", label = TRUE) + ggtitle("Seurat Cluster - pca")
v <- SCpubr::do_ViolinPlot(srat, features = c( "subsets_mito_percent"), ncol = 1, group.by = "seurat_clusters", legend.position = "none") 

d1 + d2 + v + plot_layout(ncol = 3, widths = c(2,2,4))

```

We expect up to 5 set of clusters:

- blastema cancer cells
- epithelium cancer and/or normal cells
- stroma cancer and/or normal cells
- immune cells
- endothelial cells


### Cell cycle information
```{r cell_cycle}
s.genes <- srat@assays$RNA@meta.data$gene_ids[srat@assays$RNA@meta.data$gene_symbol %in% cc.genes$s.genes]
g2m.genes <- srat@assays$RNA@meta.data$gene_ids[srat@assays$RNA@meta.data$gene_symbol %in% cc.genes$g2m.genes]

srat <- CellCycleScoring(srat, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE)
```

```{r fig.width=10, fig.height=4, out.width='100%'}
visualize_metadata(srat, meta = "Phase", group.by = "seurat_clusters")
```



```{r fig.width=8, fig.height=2.5, out.width='100%'}
visualize_metadata(srat, meta  = "S.Score", group.by = "seurat_clusters")
visualize_metadata(srat, meta = "G2M.Score", group.by = "seurat_clusters")
```


### Look at specific genes

Here, we open the table of marker genes marker-sets/CellType_metadata.csv. 
Note: we do not expect to have a clear and nice pattern of expression for all of the following markers in every tumor. 
This is just ti get a few idea.

```{r marker_genes, fig.width=12, fig.height=6, out.width='100%'}
CellType_metadata <- read_csv(file.path(module_base, "marker-sets", "CellType_metadata.csv"))

DT::datatable(CellType_metadata, caption = ("CellType_metadata"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',

                               buttons = c( 'csv', 'excel')))
```


```{r fig.width=8, fig.height=2.5, out.width='100%'}
for(feature in CellType_metadata$ENSEMBL_ID){
  if(feature %in% rownames(srat)){
    tmp <-  visualize_feature(srat, features = feature, group.by = "seurat_clusters")
    print(tmp)
  }
}
```



### Look at specific pathways

#### TP53 pathway

here we will calculate a TP53 score using AddMduleScore and the genes of the HALLMARK_P53_PATHWAY gene set.

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "H", gs_name = "HALLMARK_P53_PATHWAY", name = "TP53_score")
visualize_metadata(srat, meta = "TP53_score1", group.by = "seurat_clusters" )
```

"Normal" cells (immune, endothelial) have a slightly higher TP53 score. 


#### DNA repair pathway

here we will calculate a DNA_repair score using AddMduleScore and the genes of the HALLMARK_DNA_REPAIR gene set.

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "H", gs_name = "HALLMARK_DNA_REPAIR", name = "DNA_repair_score")
visualize_metadata(srat, meta = "DNA_repair_score1", group.by = "seurat_clusters" )
```

Note: Chemo-treated samples should have higher DNA-damage scores.


#### DROSHA target genes

```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "C3", gs_name = "DROSHA_TARGET_GENES", name = "DROSHA_score")
visualize_metadata(srat, meta = "DROSHA_score1", group.by = "seurat_clusters" )
```


#### DICER1 target genes


```{r fig.height=2.5, fig.width=8, warning=FALSE, out.width='100%'}
srat <- MSigDB_score(object = srat , category = "C3", gs_name = "DICER1_TARGET_GENES", name = "DICER1_score")
visualize_metadata(srat, meta = "DICER1_score1", group.by = "seurat_clusters" )
```


### Find marker genes for each of the seurat clusters

In addition to the list of known marker genes, we used an unbiased approach to find transcripts that characterized the different clusters.
We run DElegate::FindAllMarkers2 to find markers of the different clusters and manually check if they do make sense. 
DElegate::FindAllMarkers2 is an improved version of Seurat::FindAllMarkers based on pseudobulk differential expression method. 
Please check the preprint from Chistoph Hafemeister: https://www.biorxiv.org/content/10.1101/2023.03.28.534443v1
and tool described here: https://github.com/cancerbits/DElegate 

Of note, we won't use it for annotation, this is just here to get an idea!

```{r fig.width=10, fig.height=10, out.width='100%'}
feature_conversion <- srat@assays$RNA@meta.data
de_results   <- DElegate::FindAllMarkers2(srat, group_column = "seurat_clusters")

#filter the most relevant markers
s.markers <- de_results[de_results$padj < params$padj_thershold & de_results$log_fc > params$lfc_threshold & de_results$rate1 > params$rate1_threshold,]

# add gene symbol for easiest interpretation of the result
s.markers$gene_ids <- s.markers$feature
s.markers <- left_join(s.markers,feature_conversion, by = c( "gene_ids") )
identical(s.markers$feature, s.markers$gene_ids)     # check the quality of the merge, must be true

DT::datatable(s.markers, caption = ("marker genes"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',

                               buttons = c( 'csv', 'excel')))

# Select top 5 genes for heatmap plotting
s.markers <- na.omit(s.markers)
s.markers %>%
    group_by(group1) %>%
    top_n(n =  5, wt = log_fc) -> top5

# subset for plotting
cells <- WhichCells(srat, downsample = 100)
ss <- subset(srat, cells = cells)
ss <- ScaleData(ss, features = top5$feature)

p1 <- SCpubr::do_DimPlot(srat, reduction="umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE) + ggtitle("Seurat Cluster - umap")
p2 <- DoHeatmap(ss, features = top5$feature,  cells = cells, group.by = "seurat_clusters") + NoLegend() + 
  scale_fill_gradientn(colors =  c("#01665e","#35978f",'darkslategray3', "#f7f7f7", "#fee391","#fec44f","#F9AD03")) 
p3 <- ggplot(srat@meta.data, aes(seurat_clusters, fill = seurat_clusters)) + geom_bar() + NoLegend()


common_title <- sprintf("Unsupervised clustering %s, %d cells", srat@meta.data$orig.ident[1], ncol(srat))
show((((p1 / p3) + plot_layout(heights = c(3,2)) | p2) ) + plot_layout(widths = c(1, 2)) + plot_layout(heights = c(3,1)) + plot_annotation(title = common_title))
DT::datatable(top5[, c(1, 9, 11, 12)], caption = ("top 5 marker genes"), 
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',

                               buttons = c( 'csv', 'excel')))


```



### Enrichment analysis of marker genes

Here we perform enrichment analysis of the marker genes found in the previous section for each Seurat cluster.

We defined as universe/background all the genes expressed in the dataset, meaning the rownames of the Seurat object.

We used three gene sets from MSiGDB:

- the hallmark gene sets  are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
- the C3 : regulatory target gene sets  based on gene target predictions for microRNA seed sequences and predicted transcription factor binding sites.
- the C8 : cell type signature gene sets  curated from cluster markers identified in single-cell sequencing studies of human tissue.

We used enricher function from clusterProfiler to perform enrichment analysis.

```{r enrichment}
# define background genes = universe for enrichment
background <-  row_metadata$gene_ids

# Define gene signature per cluster
signatures <- list()
for( i in unique(s.markers$group1)){
  signatures[[paste0("cluster ",i)]] <- s.markers$feature[s.markers$group1 == i]
}
```

#### Hallmarks MSiGDB gene sets

```{r fig.width=15, fig.height=8, out.width='100%'}
tryCatch(
  expr =  {
    Enrichment_plot(category = "H", signatures = signatures, background = background)
  },
  error = function(e) {
    print("No enrichment")
  }
)
```


EMT signature should be enriched in stroma cluster.
E2F/proliferation should be enriched in blastema cluster.
MYC(N), TP53 must be enriched in blastema cluster.

#### C3 MSiGDB regulatory target gene sets 

```{r fig.width=20, fig.height=25, out.width='100%'}
tryCatch(
  expr =  {
Enrichment_plot(category = "C3", signatures = signatures, background = background)
  },
  error = function(e) {
    print("No enrichment")
  }
)
```
Here to check of we catch and MIR pattern (DROSHA; DICER1; other?)

#### C8 MSiGDB cell type signature gene sets

```{r fig.width=20, fig.height=25, out.width='100%'}
tryCatch(
  expr =  {
Enrichment_plot(category = "C8", signatures = signatures, background = background)
  },
  error = function(e) {
    print("No enrichment")
  }
)
```

The MSigDB C8 gene set is quite relevant for kidney and nephroblastoma annotations. 
Epithelial (cancer and normal) cells should be enriched in mature/adulte kidney pathways while blastema cancer cells will show enrichment of fetal kidney development pathway / cap mesenchyme. 


### Annotation from SingleR (no cancer or kidney specific dataset)

Here, we quickly checked annotations that are present in the _processed rds object. 
However, the automated annotation have not been performed using a cancer specific reference or a kidney reference. 
We do not expect a nice labelling of the cells as the overlap of cell types between the reference and the query dataset is poor. 
This support the need to do a proper label transfer from the fetal kidney atlas, which is imho the best reference that can be applied to a Wilms tumor query. 


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "singler_celltype_annotation", group.by = "seurat_clusters")
```




### Annotation from cellassign (no cancer or kidney specific dataset)


```{r fig.width=15, fig.height=10, out.width='100%'}
visualize_metadata(srat, meta = "cellassign_celltype_annotation", group.by = "seurat_clusters")
```


Even if the result of SingleR and CellAssign are not specific to a Wilms tumor dataset, it give the first impression that C12 is a cluster of endothelial cells and C11 a cluster of immune cells. 

### Azimuth annotation from fetal kidney

To be added in a next PR. Look at opened PR #706


For more information related to the reference, please go to https://www.kidneycellatlas.org/
You will find:

- interactive viewer

- h5ad files to download.

Please note that as Wilms tumor have been described to be closer to fetal kidney as mature kidney, we only used the fetal kidney atlas as the reference. 
Also check : https://www.science.org/doi/10.1126/science.aat5031

This part is imho one of the most important step that allow us to have a quick and reliable idea of the composition of the different clusters. 
The predicted compartment are defined into 4 categories:

- endothelium
- stroma
- fetal nephron 
- immune

As for SingleR and CellAssign, the annotation of immune cells and endothelial cells is straightforward. 
The stroma compartment should then contain normal and cancer stromal cells.
The fetal nephron compartment contain blastema cancer cells a well as normal and cancer epithelial cells. 

Further segregation of cancer versus normal cells will be achieved using a combination of markers/pathways (see above) and inferred CNV (to be done).



## Session Info

```{r session info}
# record the versions of the packages used in this analysis and other environment information
sessionInfo()
```
