---
title: "Using CopyKAT to identify tumor cells in Ewing sarcoma samples"
author: Ally Hawkins
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
params:
  sample_id: SCPCS000490
  library_id: SCPCL000822
  marker_gene_classification: "tumor-normal-classifications.tsv"
  no_ref_copykat_results: "copykat/no_reference"
  with_ref_copykat_results: "copykat/with_reference"
  reference_cell_file: ".reference-cells.tsv"
  results_dir: "annotate_tumor_cells_output"
---


## Introduction

This notebook explores using [`CopyKAT`](https://github.com/navinlabcode/copykat) to estimate tumor and normal cells in `r params$sample_id` from SCPCP000015. 

`CopyKAT` was run using the `run-copykat.R` script with and without a normal reference. 
These results are read into this notebook and used to: 

- Visualize diploid and aneuploid cells on the UMAP. 
- Evaluate common copy number gains and losses in Ewing sarcoma. 
- Calculate the confusion matrix comparing manual annotations of tumor cells to using `CopyKAT` to annotate tumor cells. 
- Compare the annotations from `CopyKAT` to cell type annotations from `SingleR` and `CellAssign`. 


## Setup

```{r packages}
suppressPackageStartupMessages({
  # load required packages
  library(SingleCellExperiment)
  library(ggplot2)
  library(copykat)
})

knitr::opts_chunk$set(
  dev = "jpeg"
)
```


```{r base paths}
# The base path for the OpenScPCA repository, found by its (hidden) .git directory
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

# The current data directory, found within the repository base directory
data_dir <- file.path(repository_base, "data", "current")
sample_dir <- file.path(data_dir, "SCPCP000015", params$sample_id)

# The path to this module
module_base <- file.path(repository_base, "analyses", "cell-type-ewings")
```


```{r}
# source in helper functions for make_jaccard_matrix() and jaccard()
jaccard_functions <- file.path(module_base, "scripts", "utils", "jaccard-functions.R")
source(jaccard_functions)
```


```{r paths}
# Input files
sce_filename <- glue::glue("{params$library_id}_processed.rds")
sce_file <- file.path(sample_dir, sce_filename)

obj_file <- glue::glue("{params$library_id}_final-copykat.rds")
png_file <- glue::glue("{params$library_id}_copykat_heatmap.jpeg")

copykat_objs <- c(
  no_ref = file.path(params$no_ref_copykat_results, obj_file),
  with_ref = file.path(params$with_ref_copykat_results, obj_file)
) |> 
  purrr::map(readr::read_rds)

png_list <- c(
  no_ref = file.path(params$no_ref_copykat_results, png_file),
  with_ref = file.path(params$with_ref_copykat_results, png_file)
)

# output classifications file 
copykat_results_file <- file.path(params$results_dir, glue::glue("{params$library_id}_copykat-classifications.tsv"))
```


```{r}
# read in processed sce
sce <- readr::read_rds(sce_file)

# read in tumor normal classifications 
manual_classifications_df <- readr::read_tsv(params$marker_gene_classification)

# check if marker gene annotations are present 
if(all(is.na(manual_classifications_df$marker_gene_classification))){
  has_marker_gene <- FALSE
  message("No annotations were available using only marker gene expression. 
          Any plots comparing CopyKAT to marker gene annotation will be skipped.")
} else {
  has_marker_gene <- TRUE 
}

# read in ref cells 
normal_cells <- readr::read_tsv(params$reference_cell_file) |> 
  dplyr::filter(reference_cell_class == "Normal") |> 
  dplyr::pull(barcodes)

# read in ck predictions from both reference types (no_normal and with_normal)
ck_results_df <- copykat_objs |> 
  purrr::map(\(obj){
    obj$prediction |> 
      as.data.frame()
  }) |> 
  dplyr::bind_rows(.id = "reference_used")

# read in full gene by cell copy number detection results 
full_ck_results_df <- copykat_objs |> 
  purrr::map(\(obj){
    obj$CNAmat |> 
      as.data.frame()
  }) |> 
  dplyr::bind_rows(.id = "reference_used")
```


## CopyKAT results

Below we look at the heatmaps produced by `CopyKAT`. 

### Heatmap without reference

![](`r png_list["no_ref"]`)

### Heatmap with endothelial cells as reference

![](`r png_list["with_ref"]`)9

### UMAP

Below we prepare and plot a UMAP that shows which cells are classified as diploid, aneuploid, and not defined by `CopyKAT`. 
We show a side by side UMAP with results from running `CopyKAT` both with and without a reference of normal cells. 

```{r}
umap_df <- sce |> 
  scuttle::makePerCellDF(use.dimred = "UMAP") |> 
  # replace UMAP.1 with UMAP1
  dplyr::rename_with(
        \(x) stringr::str_replace(x, "^UMAP\\.", "UMAP")
      )

cnv_df <- umap_df |> 
  # first add manual annotations
  dplyr::left_join(manual_classifications_df) |> 
  # now add copykat results
  dplyr::left_join(ck_results_df, by = c("barcodes" = "cell.names")) |> 
  dplyr::mutate(
    copykat.pred = dplyr::if_else(
      # label any reference cells that were used 
      (barcodes %in% normal_cells) & (reference_used == "with_ref"), 
      "reference", 
      copykat.pred)
  )
```

```{r}
ggplot(cnv_df, aes(x = UMAP1, y = UMAP2, color = copykat.pred)) +
  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() +
  facet_wrap(vars(reference_used))
```


### Validate common CNAs found in Ewing sarcoma

To validate some of these annotations, we can also look at some commonly found copy number variations found in Ewing sarcoma patients. 
There are a few known copy number variations in Ewing's sarcoma: 

- Gain of Chr8
- Gain of Chr12
- Gain of Chr1p
- Loss of Chr16q

Although these are the most frequent, there are patients who do not have any of these alterations and patients that only have some of these alterations. 
See [Tirode et al.,](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4264969/) and [Crompton et al.](https://doi.org/10.1158/2159-8290.CD-13-1037). 

`CopyKAT` outputs a matrix that contains the estimated copy numbers for each gene in each cell. 
We can read that in and look at the mean estimated copy numbers for each chromosome across each cell. 
We might expect that tumor cells would show an increased estimated copy number in Chr8, Chr12, and/or Chr 1 and a loss of Chr16. 

```{r}
# for every cell, calculate the mean detection level across all genes in a given chromosome
full_cnv_df <- full_ck_results_df |> 
  tidyr::pivot_longer(cols = -c(
    reference_used,
    chrom
    ),
    names_to = "barcodes",
    values_to = "cnv_detection") |> 
  dplyr::group_by(chrom, barcodes, reference_used) |> 
  dplyr::summarise(mean_cnv_detection = mean(cnv_detection)) 

# join with cnv info
cnv_df <- cnv_df |> 
  dplyr::left_join(full_cnv_df, by = c("barcodes", "reference_used")) |> 
  dplyr::filter(!is.na(chrom))
```

Let's look at the distribution of CNV estimation in cells that are called aneuploid and diploid by `CopyKAT`. 

```{r, fig.height=15, fig.width=10}
# create faceted density plots showing estimation of CNV detection across each chr of interest 
# colored by aneuploid/diploid estimation 
ggplot(cnv_df, aes(x = mean_cnv_detection, color = copykat.pred)) +
      geom_density() +
      theme_bw() +
      facet_grid(rows = vars(chrom), 
                 cols = vars(reference_used))
```


### Confusion Matrix

Below we directly compare the annotations obtained using manual classification of tumor and normal cells to annotating cells with `CopyKAT`. 
To do this, we will calculate the confusion matrix using `caret::confusionMatrix()`. 

```{r, eval=has_marker_gene}
filtered_cnv_df <- cnv_df |> 
  dplyr::filter(!(copykat.pred %in% c("not.defined", "reference")),
                # filter any low confidence calls
                !stringr::str_detect(copykat.pred, "low.conf"))
  
caret_df_list <- filtered_cnv_df |> 
  dplyr::mutate(copykat = ifelse(
    # use str_detect for test data
    copykat.pred == "diploid", "Normal", "Tumor"
  )) |> 
  # make tumor the positive class
  dplyr::mutate(
    copykat = forcats::fct_relevel(copykat, "Tumor"),
    marker_gene_classification = forcats::fct_relevel(marker_gene_classification, "Tumor")
  ) |> 
  split(cnv_df$reference_used)


caret_df_list |> 
  purrr::imap(\(df, ref_type){
    caret::confusionMatrix(
      table(
        df$marker_gene_classification, 
        df$copykat)
    )
  })
```


We can also calculate the Jaccard similarity index to visualize the amount of cells that have overlapping annotations. 

```{r, eval=has_marker_gene}
# calculate Jaccard similarity index for each reference type
jaccard_matrices <- caret_df_list |>
  purrr::map(\(df) {
    
    make_jaccard_matrix(
      df,
      "marker_gene_classification",
      "copykat.pred"
    )
  })
```

```{r, eval=has_marker_gene}
# Set heatmap padding option
heatmap_padding <- 0.2
ComplexHeatmap::ht_opt(TITLE_PADDING = grid::unit(heatmap_padding, "in"))

# heatmaps comparing tumor/normal annotations manually vs. copyKAT
heatmap <- jaccard_matrices |>
  purrr::imap(
    \(jaccard_mtx, ref_type) {
      ComplexHeatmap::Heatmap(
        t(jaccard_mtx), # transpose because matrix rows are in common & we want a vertical arrangement
        col = circlize::colorRamp2(c(0, 1), colors = c("white", "darkslateblue")),
        border = TRUE,
        ## Row parameters
        cluster_rows = TRUE,
        row_title = ref_type,
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_side = "left",
        row_names_side = "left",
        row_dend_side = "right",
        row_names_gp = grid::gpar(fontsize = 10),
        ## Column parameters
        cluster_columns = FALSE,
        column_title = "",
        column_title_gp = grid::gpar(fontsize = 12),
        column_names_side = "bottom",
        column_names_gp = grid::gpar(fontsize = 10),
        column_names_rot = 90,
        ## Legend parameters
        heatmap_legend_param = list(
          title = "Jaccard index",
          direction = "vertical",
          legend_width = unit(1.5, "in")
        ),
        show_heatmap_legend = ref_type == "no_ref",
      )
    }) |>
  # concatenate vertically into HeatmapList object
  purrr::reduce(ComplexHeatmap::`%v%`) |>
  ComplexHeatmap::draw(
    heatmap_legend_side = "right",
    # add a margin to the heatmap so labels don't get cut off
    padding = unit(c(2, 20, 2, 2), "mm")
  )
```


### Compare CopyKAT to SingleR and CellAssign

Lastly, we will compare the annotations from `CopyKAT` to those obtained using `SingleR` and `CellAssign` by calculating the Jaccard similarity index. 
For this comparison we will use just the annotations from `CopyKAT` with no reference. 

```{r}
celltype_columns <- c(
  "singler_celltype_annotation",
  "cellassign_celltype_annotation"
)

# filter to only get annotations from no ref
no_ref_only <- cnv_df |> 
  dplyr::filter(reference_used == "no_ref")

# create jaccard matrices for SingleR and CellAssign compared to aneuploid/diploid 
jaccard_matrices <- celltype_columns |>
  purrr::map(\(name) {
    make_jaccard_matrix(
      no_ref_only,
      "copykat.pred",
      name
    )
  }) |> 
  purrr::set_names("SingleR", "CellAssign")
```


```{r, fig.height=10, fig.width=7}
# Set heatmap padding option
heatmap_padding <- 0.2
ComplexHeatmap::ht_opt(TITLE_PADDING = grid::unit(heatmap_padding, "in"))

# list of heatmaps looking at SingleR/ CellAssign vs tumor/normal 
heatmap <- jaccard_matrices |>
  purrr::imap(
    \(celltype_mat, celltype_method) {
      ComplexHeatmap::Heatmap(
        t(celltype_mat), # transpose because matrix rows are in common & we want a vertical arrangement
        col = circlize::colorRamp2(c(0, 1), colors = c("white", "darkslateblue")),
        border = TRUE,
        ## Row parameters
        cluster_rows = TRUE,
        row_title = celltype_method,
        row_title_gp = grid::gpar(fontsize = 12),
        row_title_side = "left",
        row_names_side = "left",
        row_dend_side = "right",
        row_names_gp = grid::gpar(fontsize = 10),
        ## Column parameters
        cluster_columns = FALSE,
        column_title = "",
        column_title_gp = grid::gpar(fontsize = 12),
        column_names_side = "bottom",
        column_names_gp = grid::gpar(fontsize = 10),
        column_names_rot = 90,
        ## Legend parameters
        heatmap_legend_param = list(
          title = "Jaccard index",
          direction = "vertical",
          legend_width = unit(1.5, "in")
        ),
        show_heatmap_legend = celltype_method == "SingleR",
      )
    }) |>
  # concatenate vertically into HeatmapList object
  purrr::reduce(ComplexHeatmap::`%v%`) |>
  ComplexHeatmap::draw(
    heatmap_legend_side = "right",
    # add a margin to the heatmap so labels don't get cut off
    padding = unit(c(2, 20, 2, 2), "mm")
  )
```

## Save annotations 

```{r}
celltype_df <- cnv_df |> 
  dplyr::select(barcodes, 
                reference_used,
                copykat.pred) |> 
  dplyr::distinct() |> 
  tidyr::pivot_wider(
    names_from = reference_used,
    values_from = copykat.pred
  )

all_info_df <- cnv_df |> 
  dplyr::select(barcodes,
                mean_cnv_detection) |> 
  dplyr::group_by(barcodes) |>
  dplyr::mutate(mean_cnv_detection = mean(mean_cnv_detection)) |> 
  dplyr::distinct() |> 
  dplyr::left_join(celltype_df)

readr::write_tsv(all_info_df, copykat_results_file)
```


## Session Info

```{r session info}
# record the versions of the packages used in this analysis and other environment information
sessionInfo()
```
